sympy/solvers/ode/single.py

```diff
@@ -675,7 +676,8 @@ def _matches(self):
         self.eqs = []
         eq = eq.collect(f(x), func = cancel)
         eq = fraction(factor(eq))[0]
-        roots = factor_list(eq)[1]
+
+        roots = [g.args if g.func.is_Pow else (g,1) for g in list(Mul.make_args(factor(eq))) if g.args!=x.args ]
```

Comment:
I understand that you got here incrementally but this line of code is too much I think for a list comprehension.

Okay so `g.args if g.func.is_Pow else (g,1)` is basically `g.as_base_exp()` (although it will handle `exp(x)` differently).

So
```
factors = Mul.make_args(factor(eq))
roots = [fac.as_base_exp() for fac in factors if g.args != x.args]
```
Isn't `x.args` just an empty tuple?

Reply:
Didn't know about `as_base_exp`. I'll replace `x.args` with `len(g.args)!=0`, I think it's nicer...

Reply:
Why is the `len(g.args)` check needed?

Reply:
g.args can be empty and a couple of lines down, roots is expected to have entries of the form [(a,b),(c,d)] so we prevent empty () entries in the list in this way

Reply:
Does `as_base_exp` take care of the empty `()` args case?

Reply:
as_base_exp(1) will return (1,1). Without filtering, roots could contain (5,1) when eq=5*x. 4 lines down, we check if roots 'has' the function f(x), and .has fails on int.

```python
    for base,expo in roots:
        if base.has(f(x)):
```